### PR TITLE
Create pure_call.md

### PR DESCRIPTION
**Pull request by vbuterin**
At `0xf5`, add an opcode `PURE_CALL` that takes seven arguments from the stack: GAS, CODESTART, CODELENGTH, DATASTART, DATALENGTH, OUTPUTSTART, OUTPUTLENGTH. Processes a call with the given code and data that cannot read or write state, and pastes the result into the output memory slice.

### CODE REVIEW

**Comment by gcolvin** [comment]
Does that mean it can only make pure calls itself?

**Comment by vbuterin** [comment]
> Does that mean it can only make pure calls itself?

Not sure I understand; what do you mean?

The EIP as written does not allow the child context to make *any* calls except for pure precompiles in order to guarantee state independence. I suppose you could allow the child context to itself make PURE_CALLs if you really want to, though there isn't *too* much additional value gained from that.

**Comment by gcolvin** [comment]
Given how tiny contracts are, it provides a safe way to break a program into small, safe pieces, with all state interaction carefully isolated.

**Comment by vbuterin** [comment]
Ok. I have no problem with allowing pure calls from inside a pure call context then.

File: EIPS/eip_195_pure_call.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+### Parameters
+
+* BASE_GAS_COST: 80
+* CODE_BYTES_PER_GAS: 15
```
Comment (chriseth) [inline EIPS/eip_195_pure_call.md]: Assuming we want to switch to EVM 1.5 - is that really the amount of gas we want to charge? Or shall we completely revise the scheme with 1.5?

Comment (gcolvin) [inline EIPS/eip_195_pure_call.md]: What revisions do you have in mind?

Comment (chriseth) [inline EIPS/eip_195_pure_call.md]: Perhaps to add some more explanation: In EVM1.5, code that is about to be executed needs to be analyzed and sightly processed first. For regular contracts, this is done once at creation time and we can use the gas from the code deposit to pay for this preprocessing. Here, 15 bytes per gas could be too low for that.

Comment (gcolvin) [inline EIPS/eip_195_pure_call.md]: Agreed.  Not sure what the right amount is, as the preprocessing can run from simple validation to full compilation.

**Comment by Arachnid** [comment]
Please make compliant with EIP-0 (eg, add a header with number/status etc), renumber to EIP 195, and update in response to any comments, and tag with `editor-needs-to-review` when it's ready for merge.

**Comment by obscuren** [comment]
What will `ADDRESS` return?

**Comment by chriseth** [comment]
I would say that `address` is also part of the state, so it should fail.

**Comment by holiman** [comment]
The EIP says it may make calls to non-state-changing precompiles. For forward compatibility; I'd like to clarify so it's clear that any non-state-changing precompiles that are added in the future are automatically included in the set of allowed precompiles.

**Comment by obscuren** [comment]
@chriseth yes, actually all auxiliary data including state read/write any calls with the exemption of non-state altering pre compiles will throw. Auxiliary includes address, coinbase, time, difficulty, blockheader, gaslimit

**Comment by holiman** [comment]
Also...

- `CODESIZE` returns `CODELENGTH`?
- 0-padding. Normally, `CODECOPY` and `DATACOPY` pads with zeroes when reading out of bounds. I presume it would do so here aswell.
- How about making `CODE_BYTES_PER_GAS` to `16`, to make it into division by 16 instead of 15 ?

**Comment by axic** [comment]
> `CODESIZE` returns `CODELENGTH`?

Isn't that a state reading operation and therefore forbidden?

> The EIP says it may make calls to non-state-changing precompiles. 

Is that actually a good idea? I think with time that might result in a very complex list of exceptions for `CALL` & co.

> Ok. I have no problem with allowing pure calls from inside a pure call context; we can do that.

Also since `PURE_CALL` takes a memory location for code as opposed to an account, how can it make calls to other *pure functions*?

**Comment by holiman** [comment]
>> The EIP says it may make calls to non-state-changing precompiles.
> Is that actually a good idea? I think with time that might result in a very complex list of exceptions for CALL & co.

Well, if I remember correctly, all precompiles (including metropolis) are non-state changing. So it would be all precompiles, rather than a list of exceptions. I do see a point in pure calls being able to use e.g. bigint math and ecrecover. 

> Also since PURE_CALL takes a memory location for code as opposed to an account, the only way it can make calls to other pure functions is by embedding the bytecode and storing it to memory.

Yes, asides from the precompiles. So what is currently suggested in the EIP, afaict, is 

* Allow `CALL` to precompiles
* Allow `PURE_CALL` (and thus provide the code along with it)

**Comment by Arachnid** [comment]
Please update to follow the template [here](https://github.com/ethereum/EIPs/blob/master/eip-X.md) and rename to `EIP-195.md`, and we'll merge as draft.

File: EIPS/eip_195_pure_call.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+Fail if `input_gas < gas_cost` or if `CODELENGTH > 24000`; otherwise execute a call with code equal to the memory slice `CODESTART...CODESTART + CODELENGTH - 1`, input data equal to the memory slice `DATASTART...DATASTART + DATALENGTH - 1`, gas equal to `min(GAS, (input_gas - gas_cost) - floor((input_gas - gas_cost) / 64))` and value 0. The inner context may NOT read state, write state or make any sub-calls except to non-state-changing precompiles; any attempt to do so throws an exception. If execution succeeds, fill the memory slice `OUTPUTSTART...OUTPUTSTART + OUTPUTLENGTH - 1` with the return data with similar rules to how this is handled in CALL, CALLCODE and DELEGATECALL, and push 1 onto the stack. If execution fails, push 0 onto the stack.
+
+### Rationale
+
+This is useful for:
+
+* Adding functional programming features for higher-level languages
+* Generally allowing for forms of computation that are safer and can be efficiently optimized/memoized due to their state-independence
+* Processing validation code for Casper validators in the future
```
Comment (pirapira) [inline EIPS/eip_195_pure_call.md]: Please waive the copyright with CC0.

**Comment by axic** [comment]
I think two versions of this instead the current proposed one would make more sense:
1) Would read from code offset and length within the current contracts code. This means calling a function part of the callee contract wouldn't need to do a `codecopy` to memory first in order to pass it to `purecall`. 
2) Load code from a destination address and execute it with restricting both state reading and writing operations (`staticcall`'s counterpart). For rules on what are state reading operations one could refer to Solidity's categorisation: https://github.com/ethereum/solidity/blob/develop/libevmasm/SemanticInformation.cpp#L192-L218

Though for 2) practically it would need a new opcode to determine the current context type (`ISPURE` or a more generic approach is `CALLTYPE` to return `regular`, `delegate`, `static`, `pure`). At least Solidity checks that no value was sent in a call if the function doesn't expect it and that is done via the `CALLVALUE` opcode, which is restricted in the pure context.

**Comment by Arachnid** [comment]
This is a courtesy notice to let you know that the format for EIPs has been modified slightly. If you want your draft merged, you will need to make some small changes to how your EIP is formatted:

 - Frontmatter is now contained between lines with only a triple dash ('---')
 - Headers in the frontmatter are now lowercase.

If your PR is editing an existing EIP rather than creating a new one, this has already been done for you, and you need only rebase your PR.

In addition, a continuous build has been setup, which will check your PR against the rules for EIP formatting automatically once you update your PR. This build ensures all required headers are present, as well as performing a number of other checks.

Please rebase your PR against the latest master, and edit your PR to use the above format for frontmatter. For convenience, here's a sample header you can copy and adapt:

```
---
eip: <num>
title: <title>
author: <author>
type: [Standards Track|Informational|Meta]
category: [Core|Networking|Interface|ERC] (for type: Standards Track only)
status: Draft
created: <date>
---
```

**Comment by fubuloubu** [comment]
@pipermerriam we should revisit this proposal, specifically @axic's last comment, with @vbuterin in context of Casper

**Comment by fulldecent** [comment]
I propose a different PURECALL that is more like STATICCALL:

* Use the same arguments as STATICCALL: gas, to, in offset, in size, out offset, out size.
* Update Yellow Paper "This states that the execution is in an exceptional halting-state if..." to include "... state access is attempted during a pure call."

---

I do not like the proposed CODESTART, CODELENGTH parameters proposed above because that is effectively a JUMP without a JUMPDEST. Also the semantics of CALL are well understood. If we wanted the seven-argument opcode as above then I would call the GOTOPURE instead.

**Comment by MicahZoltu** [comment]
@vbuterin Do you still want to pursue this?  If so it will need to be updated to match the latest EIP formatting requirements.

**Pull request closed without merging**